firebase-ai/src/main/kotlin/com/google/firebase/ai/common/APIController.kt

```diff
@@ -0,0 +1,360 @@
+   * @param config An optional [HttpRequestBuilder] callback for request configuration.
+   * @return A [Flow] of response objects of type [R].
+   */
+  private inline fun <reified R : Response> HttpClient.postStream(
+    url: String,
+    crossinline config: HttpRequestBuilder.() -> Unit = {},
+  ): Flow<R> = channelFlow {
+    launch(CoroutineName("postStream")) {
+      preparePost(url) {
+          applyHeaderProvider()
+          config()
+        }
+        .execute {
+          validateResponse(it)
+
+          val channel = it.bodyAsChannel()
+          val flow = JSON.decodeToFlow<R>(channel)
+
+          flow.collect { send(it) }
+        }
+    }
+  }
+
+  companion object {
+    private val TAG = APIController::class.java.simpleName
+
+    private fun getVersionNumber(app: FirebaseApp): Int {
+      try {
+        val context = app.applicationContext
+        return context.packageManager.getPackageInfo(context.packageName, 0).versionCode
+      } catch (e: Exception) {
+        Log.d(TAG, "Error while getting app version: ${e.message}")
+        return 0
+      }
+    }
+  }
+}
+
+internal interface HeaderProvider {
+  val timeout: Duration
+
+  suspend fun generateHeaders(): Map<String, String>
+}
+
+private suspend fun validateResponse(response: HttpResponse) {
+  if (response.status == HttpStatusCode.OK) return
+
+  val htmlContentType = ContentType.Text.Html.withCharset(Charset.forName("utf-8"))
+  if (response.status == HttpStatusCode.NotFound && response.contentType() == htmlContentType)
+    throw ServerException(
+      """URL not found. Please verify the location used to create the `FirebaseAI` object
+          | See https://cloud.google.com/vertex-ai/generative-ai/docs/learn/locations#available-regions
+          | for the list of available locations. Raw response: ${response.bodyAsText()}"""
+        .trimMargin()
+    )
+  val text = response.bodyAsText()
+  val error =
+    try {
+      JSON.decodeFromString<GRpcErrorResponse>(text).error
+    } catch (e: Throwable) {
+      throw ServerException("Unexpected Response:\n$text $e")
+    }
+  val message = error.message
+  if (message.contains("API key not valid")) {
+    throw InvalidAPIKeyException(message)
+  }
+  // TODO (b/325117891): Use a better method than string matching.
+  if (message == "User location is not supported for the API use.") {
+    throw UnsupportedUserLocationException()
+  }
+  if (message.contains("quota")) {
+    throw QuotaExceededException(message)
+  }
+  if (message.contains("The prompt could not be submitted")) {
+    throw PromptBlockedException(message)
+  }
+  getServiceDisabledErrorDetailsOrNull(error)?.let {
+    val errorMessage =
+      if (it.metadata?.get("service") == "firebasevertexai.googleapis.com") {
+        """
+        The Firebase AI SDK requires the Vertex AI in Firebase API
+        (`firebasevertexai.googleapis.com`) to be enabled in your Firebase project. Enable this API
```

Comment:
Is this still the name of the service?

Reply:
That's a great question!

Reply:
This error only pops up if the service *is* firebasevertexai.googleapis.com, so I think this error should be good

Reply:
That's a good point. But is there another `firebaseai` adjacent service that we should be throwing a similar message for?